### DISCUSSION TITLE
What is this error:  EROFS: read-only file system, open? And how do I fix it

### DISCUSSION THREAD

**Discussion by stijnwesterhof** [vercel/community#314] [Help] [answered]
Since this week this error popped up in my error logging:

> INFO	error:  EROFS: read-only file system, open 'store/initialCountries.json'

I have a node script that I run before I do `npm run build` which fetches some data from an API and stores it in an .json file. This file is being used during runtime on basically every page we have.

I never had this error before and I didn't change anything related to that code. Also, the json file is present during runtime and our pages have the correct data. 

This error is causing a lot of Errored GB-hours of serverless function execution and I would like to get rid of it, but I have no idea what I'm doing wrong.

Another thing I don't understand is that it's also erroring on image routes (?)
![image](https://user-images.githubusercontent.com/15048777/152367575-f52f77a8-2b4c-4605-a8d2-31f8897e4116.png)

**Comment by samcx** [comment]
@stijnwesterhof Apologies for the late response. Can you confirm if you are still seeing this error?

**Reply by amartuvshin-bold** [reply]
Hi. @samsisle 

My error logging: 
EROFS: read-only file system, open 'public/data/attorneys.json'

How can I fix it?

**Reply by amartuvshin-bold** [reply]
I've deployed my project to Vercel. It deploys fine

**Reply by samcx** [reply]
Tough to say what is going on here. Is it possible to share with us a public, reproducible repository so we can take a look?

**Reply by liftyourgame** [reply]
I had this error. Fixed by writing the file under /tmp

**Reply by stijnwesterhof** [reply]
@liftyourgame Is this a folder you need to create yourself? Or is this a folder that is somewhere in the build folder?

**Reply by liftyourgame** [reply]
Most containers have a /tmp folder. It's a UNIX convention. Usually held in memory and cleared on reboot. Don't need to create the folder yourself.

**Comment by QGB** [comment]
vercel "Read-only file system"

**Comment by ThakurSaad** [comment]
EROFS: read-only file system, open 'users.json'
I had the same error

**Comment by amyegan** [comment, marked as answer]
This error is most often caused by attempts to write data from a serverless function to a read-only file at runtime. You can work around this by using `/tmp` if you only need **temporary** storage, as mentioned in [an another comment thread](https://github.com/orgs/vercel/discussions/314#discussioncomment-2691930). If you need persistent storage, then a more permanent data storage service is needed. There are some [database integration](https://vercel.com/integrations#databases) options that can make it pretty easy to set up!

**Reply by ThakurSaad** [reply]
@amyegan you are correct about my problem. I am trying to write data in a **.json** file. 
`fs.writeFileSync(__dirname + "/../users.json", JSON.stringify(users))`. In this case, how should I use `/tmp`

**Reply by amyegan** [reply]
These examples might help: https://stackoverflow.com/a/53770877 and https://stackoverflow.com/a/53762838

**Comment by heymage** [comment]
Just to clarify it for other users: It really just has to be `/tmp`, not `./tmp`. Not a big deal but I had to redeploy my preview some times to get this. Hope that saves someone some time :)

**Comment by ThakurSaad** [comment]
I am trying to write data in a `.json` file.
`fs.writeFileSync(__dirname + "/../users.json", JSON.stringify(users))` In this case, how should I use `/tmp`

**Reply by heymage** [reply]
I guess you can just use `fs.writeFileSync("/tmp/users.json", JSON.stringify(users))`.

**Comment by irwansyafani** [comment]
Alternative answer, if you put `/tmp` and still not working you may to have tell the path to go out the current location first. I fixed my issue with this way cause I when I put `/tmp/<file.ext>` its still not working

> before: `/tmp/file.png`
> after: `/../../../../../tmp/file.png`

you may ask where I know that I have to go out 5 times (`../`)? I check it through **View Function Logs inside** Vercel project

**Reply by amyegan** [reply]
Hi @irwansyafani. You shouldn't need to back out multiple levels to write to the `tmp` directory. I have a serverless function working with this setup:
```
 const filePath = path.join("/tmp", "data.json");
 fs.writeFileSync(filePath, JSON.stringify(data));
```
Are you able to share an example so I can look into why this isn't working for you?
